Add menu for queued experiments

giwon2004 opened the pull request:
### This is a new PR made from the contaminated pull request #102.

I made a `QMenu` object that displays a menu when right-clicked:
![image](https://github.com/snu-quiqcl/iquip/assets/66408173/f40e6c55-5374-490b-a1dd-19c7bc18e04e)
(The space in front of the text will be deleted in the future, using `QMenu.setStyleSheet`.)

Here, Edit will be future implemented, and the other two are connected to POST request mentioned in [artiq-proxy#44](https://github.com/snu-quiqcl/artiq-proxy/issues/44) and #96.

This closes #96.

giwon2004 committed:
bring commits in one whole commit

giwon2004 commented:
The pylint & unit tests' errors are not caused by this commit.

giwon2004 committed:
change unit test to match change in the main code

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -30,6 +34,24 @@ def _dismiss_items(layout: Optional[QLayout] = None):
+          - worker.done: the signal that is emitted when the work is done.
+    """
+    thread = QThread(parent=parent)
+    worker = SchedulerPostWorker("delete")
```
Why is `worker` overwritten here?

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -59,6 +81,22 @@ def __init__(self, parent: Optional[QWidget] = None):
+    """Customized QListView class to detect right-click input.
+    
+    Signals:
+        rightButtonPressed(QMouseEvent): The information of the click input is sent.
```
In the docstring for signals, we use intuitive argument names instead of their types, since the types are already in the codes right below.
For example, `rightButtonPressed(mouseEvent)`.

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -59,6 +81,22 @@ def __init__(self, parent: Optional[QWidget] = None):
+        if event.type() == QEvent.MouseButtonPress:
+            if event.button() == Qt.RightButton:
+                self.rightButtonPressed.emit(event)
+        super().mousePressEvent(event) # hand the signal to drag & drop
```
Please make 2 spaces before comments (`__#` instead of `_#`).

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -59,6 +81,22 @@ def __init__(self, parent: Optional[QWidget] = None):
+
+    def mousePressEvent(self, event):
+        """Overridden."""
+        if event.type() == QEvent.MouseButtonPress:
+            if event.button() == Qt.RightButton:
```
Why don't you use `and` instead of nesting 2 `if` statements? Is it for future use?

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -223,6 +288,33 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+            index = self.schedulerFrame.model.index(i)
+            if self.schedulerFrame.queueView.rectForIndex(index).contains(event.pos()):
+                rid = self.schedulerFrame.model.data(index).rid
+                menu = QMenu(self.schedulerFrame)
+                edit = menu.addAction("Edit")
+                delete = menu.addAction("Delete")
+                request_termination = menu.addAction("Request termination")
```
Is it common to create menu instances every time? Is it hard to reuse a single `QMenu` instance?

kangz12345 commented:
Please resolve the pylint warning.

giwon2004 committed:
reuse `QMenu` instance

giwon2004 committed:
Delete whitespaces around `=` of keyword arguments

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -282,13 +280,22 @@ class SchedulerApp(qiwis.BaseApp):
         self.schedulerFrame.queueView.rightButtonPressed.connect(self.displayMenu)
+        self.menu = QMenu(self.schedulerFrame)
+        # TODO(giwon2004) Remove icon space from the menu list (use menu.setStyleSheet)
+        self.signals = {
```
Why is the name "signals", not "actions"?

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -30,6 +34,23 @@ def _dismiss_items(layout: Optional[QLayout] = None):
                 _dismiss_items(item.layout())
 
 
+def _run_thread_with_worker(worker: QObject, parent: Optional[QObject] = None):
+    """Runs another thread with given worker.
```
The name seems awkward since it does not actually run the thread, but create and prepare the threaded worker.

giwon2004 committed:
change namings

kangz12345 commented on iquip/apps/scheduler.py:
```diff
@@ -34,7 +34,7 @@ def _dismiss_items(layout: Optional[QLayout] = None):
 
-def _run_thread_with_worker(worker: QObject, parent: Optional[QObject] = None):
+def _thread_with_worker(worker: QObject, parent: Optional[QObject] = None):
     """Runs another thread with given worker.
```
Good, now please annotate the return type: `def _thread_with_worker(...) -> QThread:`.
Also, please update the docstring summary as well: `"""Returns a new thread with given worker.`

kangz12345 reviewed (changes requested):
Do you know why the pylint and unit test actions fail?

giwon2004 commented:
> Do you know why the pylint and unit test actions fail?

We discussed about this in the previous meeting, it is about #112. I made this branch from `develop` after the commit that made the errors.

giwon2004 committed:
edit docstring

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -30,6 +34,23 @@ def _dismiss_items(layout: Optional[QLayout] = None):
+    """Returns a new thread with given worker.
+
+    Args:
+        worker: The worker that must be run through another thread. It must have:
```
Please add a description about `parent`.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -59,6 +80,21 @@ def __init__(self, parent: Optional[QWidget] = None):
 
 
+class ExperimentListView(QListView):
+    """Customized QListView class to detect right-click input.
```
I think this description is too minor.

How about writing that this list view contains the running experiments.

BECATRUE replied:
Please check it! @giwon2004

giwon2004 replied:
I did this under this line! Github only shows the modification on that exact line :(

BECATRUE replied:
Oh, you added a description under this line!
I'm sorry I didn't check well.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -59,6 +80,21 @@ def __init__(self, parent: Optional[QWidget] = None):
+    """
+    rightButtonPressed = pyqtSignal(QMouseEvent)
+
+    def mousePressEvent(self, event):
```
Can you add the type hinting about `event`?

BECATRUE commented on tests/test_scheduler.py:
```diff
@@ -123,8 +119,8 @@ def test_modify_experiment(self):
         experiment_delete = SubmittedExperimentInfo(rid=2, priority=1)
-        app.schedulerFrame.model.experimentQueue.append(experiment_delete)
-        app.schedulerFrame.model.experimentQueue.append(experiment_change)
+        app.addExperiment(experiment_delete)
```
Please modify the test function name!
I think it should be updated to `test_add_experiment()`.

giwon2004 replied:
I do not get it, this part tests `app.changeExperiment` too, as a functional test.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -212,17 +245,77 @@ def sizeHint(self, option: QStyleOptionViewItem, index: QModelIndex) -> QSize:
+            "edit": self.menu.addAction("Edit"),
+            "delete": self.menu.addAction("Delete"),
+            "terminate": self.menu.addAction("Request termination")
+        }
```
👍

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -212,17 +245,77 @@ def sizeHint(self, option: QStyleOptionViewItem, index: QModelIndex) -> QSize:
+        """
+        for i in range(self.schedulerFrame.model.rowCount()):
+            index = self.schedulerFrame.model.index(i)
+            if self.schedulerFrame.queueView.rectForIndex(index).contains(event.pos()):
```
Oh.. Is it the best solution?

Sure, this code seems to work well so you don't have to modify it now, but it looks little convoluted.

giwon2004 replied:
True, but I could not find the function that directly gives the index :(

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -212,17 +245,77 @@ def sizeHint(self, option: QStyleOptionViewItem, index: QModelIndex) -> QSize:
+                # TODO(giwon2004) Create an app for editing scannables.
+                    pass
+                elif action == self.actions["delete"]:
+                    _thread_with_worker(SchedulerPostWorker("delete", rid), self).start()
```
Unless the `QThread` object isn't stored as an attribute of this class, does it work well?

I'm worried becuase the thread object may be removed before it finishes because the thread is defined only in `displayMenu()`.

giwon2004 replied:
I have tested the thread object, and the termination of `displayMenu` did not delete it.
I guess it works well as the thread has a parent `self`.

However, I found a new problem: it does not run `worker.run` here! What could be the problem?

BECATRUE replied:
> I have tested the thread object, and the termination of displayMenu did not delete it.
I guess it works well as the thread has a parent self.

Good!

BECATRUE replied:
> However, I found a new problem: it does not run worker.run here! What could be the problem?

What do you mean?
Even if the thread starts, isn't `worker.run()` called?

giwon2004 replied:
> Even if the thread starts, isn't worker.run() called?

No, it does not call the function. :(

BECATRUE replied:
Umm.. I think the problem should be resolved in this PR..
I also try to find out the reasons, so please hold for a moment.

BECATRUE replied:
How did you check the problem?
In this branch, fetching the actual list from the proxy server isn't implemented yet, so I can't open the menu.

giwon2004 replied:
I made a dummy experiment by manually adding an experiment by `self.addExperiment`. 
Then I used right-click on it and clicked Edit.

Here I made the thread have nothing to do (command-ize the `request` part), but the error message `QThread destroyed while running` showed up after I quit the app while it did not show when the function `_thread_with_worker` is not called.

This showed the thread does not end when the call itself dies, but also showed the thread does not end _until the end_.

BECATRUE commented on iquip/apps/scheduler.py:
```diff
@@ -212,17 +245,77 @@ def sizeHint(self, option: QStyleOptionViewItem, index: QModelIndex) -> QSize:
+    def run(self):
+        """Overridden."""
+        basePath = "http://127.0.0.1:8000/experiment/"
+        requests.post(basePath + self.mode, params={"rid": self.rid}, timeout=10)
```
How simple it is!

BECATRUE reviewed (changes requested):
Did you check if the thread works well and doesn't block the main event loop?

giwon2004 committed:
docstring change

It is not running; `worker.run` is not called.

giwon2004 committed:
make `worker` survive

I am not sure whether calling the same `worker` again and again. I need to check it later.

giwon2004 committed:
Add `pyqtSlot` decorator

giwon2004 commented:
I am not sure whether using the same `self.worker` works when the menu is clicked twice, and this needs a test.

BECATRUE commented:
> I am not sure whether using the same `self.worker` works when the menu is clicked twice, and this needs a test.

If so, it may be a simple solution that makes the menu or the button disable until `self.worker` finishes.

BECATRUE reviewed (approved):
LGTM!

giwon2004 commented:
> If so, it may be a simple solution that makes the menu or the button disable until `self.worker` finishes.

No, this is about the case when the menu thread itself is called twice. (Making a new thread with an old worker)
